2.0-proposal/RELATIONS.md

```diff
@@ -0,0 +1,248 @@
+added to an application are automatically added to all of that application's
+units, even ones that join after the fact, and if a flag is added to at least
+one of an application's units, that application will be considered to have that
+flag.
+
+You can filter the list of applications and units by flags.  For example:
+
+```python
+class MyRelation(SimpleRelation):
+    class flags:
+        changed = 'relation.{relation_name}.changed'
+        ready = 'relation.{relation_name}.ready'
+
+    @when(flags.changed)
+    def changed(self):
+        for app in self.applications:
+            for unit in app.units.with_flag(self.flags.changed):
+                host = unit.data.host
+                print('unit {} host changed: {}'.format(unit.name, host))
+                if host:
+                    unit.add_flag(self.flags.ready)
+                else:
+                    unit.remove_flag(self.flags.ready)
+                # or unit.toggle_flag(self.flags.ready, host)
+
+        for app in self.applications.without_flag(self.flags.ready):
+            print('Waiting on {}'.format(app.name))
+
+    def get_hosts(self):
+        for app in self.applications.with_flag(self.flags.ready):
+            for unit in app.units.with_flag(self.flags.ready):
+                yield unit.data.host
+```
+
+In the relatively common case that you don't expect or care about multiple
+related applications, you can use the `self.all_units` view, which is roughly
+analogous to `[u for a in self.applications for u in a.units]` except that it
+also supports the filter and flag setting / removing methods.
+
+It's important to note that each application in the collection actually
+represents a relation, and so a given remote application might actually appear
+in the collection multiple times if it is related more than once (for example,
+if an application needed multiple databases).  However, the collection of units
+and flags associated with the application or units will be kept separate.  You
+can distinguish between the application instances by their `relation_id`
+property.
+
+## Relation Data
+
+As shown above, each unit has a data attribute.  This behaves like a
+read-only `defaultdict(None)` which allows the keys to be accessed as
+attributes (with dashes converted to underscores).  The values are always
+strings or `None`.
+
+There is also a helper view for the relation data set by remote units, in case
+you expect only the leader to be providing data or for the remote units to
+agree on (some subset) of the data.  For example:
+
+```python
+class MySQLClient(SimpleRelation):
+    class flags:
+        changed = 'relation.{relation_name}.changed'
+        ready = 'relation.{relation_name}.ready'
+
+    @when(flags.changed)
+    def changed(self):
+        data = self.all_units.merged_data
+        # we assume data.host to be the master
+        if all([data.host, data.database, data.user, data.password]):
+            self.add_flag(self.flags.ready)
```

Comment:
Yeah, I was thinking that it might be useful to have flags apply to different levels of the relation, with flags applied to units only applying to that unit, flags applied to applications (relation IDs) automatically applying to all units of that application (even new ones that join later), and flags applied to the relation itself automatically applying to all units of all applications.  But reading that now, that's just going to lead to confusion and we should make it so flags only ever apply to units.  We can then have helpers on the views that make it easy to add / remove flags from sets of units, so this would become: `self.all_units.add_flag(self.flags.ready)`